Code/Source/solver/mat_fun.h

```diff
@@ -61,20 +115,307 @@ namespace mat_fun {
+        };
+
+        // Return the double dot product
+        return A.contract(B, contractionDims);
+
+        // // Initialize the result tensor
+        // Eigen::TensorFixedSize<double, Eigen::Sizes<nsd, nsd, nsd, nsd>> C;
+
+        // // Compute the double dot product: C_ijkl = A_ijmn * B_kmln
+        // for (int i = 0; i < nsd; ++i) {
+        //     for (int j = 0; j < nsd; ++j) {
+        //         for (int k = 0; k < nsd; ++k) {
+        //             for (int l = 0; l < nsd; ++l) {
+        //                 C(i,j,k,l) = 0.0;
+        //                 for (int m = 0; m < nsd; ++m) {
+        //                     for (int n = 0; n < nsd; ++n) {
+        //                         C(i,j,k,l) += A(i,j,m,n) * B(k,m,l,n);
+        //                     }
+        //                 }
+        //             }
+        //         }
+        //     }
+        // }
+
+        // return C;
+    }
+
     Tensor4<double> ten_dyad_prod(const Array<double>& A, const Array<double>& B, const int nd);
+    
+    /**
+     * @brief Compute the dyadic product of two 2nd order tensors A and B, C_ijkl = A_ij * B_kl
+     * 
+     * @tparam nsd, the number of spatial dimensions
+     * @param A, the first 2nd order tensor
+     * @param B, the second 2nd order tensor
+     * @return Eigen::TensorFixedSize<double, Eigen::Sizes<nsd, nsd, nsd, nsd>> 
+     */
+    template <int nsd>
+    Eigen::TensorFixedSize<double, Eigen::Sizes<nsd, nsd, nsd, nsd>> 
+    ten_dyad_prod_eigen(const Eigen::Matrix<double, nsd, nsd>& A, const Eigen::Matrix<double, nsd, nsd>& B) {
+       
+        // // Create empty contraction dimensions
+        // Eigen::array<Eigen::IndexPair<int>, 0> contractionDims = {};
+
+        // // Convert the input matrices to Eigen::TensorFixedSize so we can use the contract method
+        // Eigen::TensorFixedSize<double, Eigen::Sizes<nsd, nsd>> A_tensor = Eigen::TensorMap<const Eigen::Tensor<const double, 2>>(A.data(), nsd, nsd);
+        // Eigen::TensorFixedSize<double, Eigen::Sizes<nsd, nsd>> B_tensor = Eigen::TensorMap<const Eigen::Tensor<const double, 2>>(B.data(), nsd, nsd);
+
+        // // The contraction of A and B with no dimensions specified is the outer product
+        // // of the two matrices, C_ijkl = A_ij * B_kl
+        // return A_tensor.contract(B_tensor, contractionDims);
+
+        // For some reason, the Eigen::Tensor contract function above is slower than the manual implementation below
+        
+        // Initialize the result tensor
+        Eigen::TensorFixedSize<double, Eigen::Sizes<nsd, nsd, nsd, nsd>> C;
+
+        // Compute the dyadic product: C_ijkl = A_ij * B_kl
+        for (int i = 0; i < nsd; ++i) {
+            for (int j = 0; j < nsd; ++j) {
+                for (int k = 0; k < nsd; ++k) {
+                    for (int l = 0; l < nsd; ++l) {
+                        C(i,j,k,l) = A(i,j) * B(k,l);
+                    }
+                }
+            }
+        }
```

Comment:
Yes, the for loop is faster in this case. This may be because the tensors we are dealing with are small (2x2 or 3x3), so the contract method is overkill.